Zend/zend_smart_str.c

```diff
@@ -169,3 +169,30 @@ ZEND_API void ZEND_FASTCALL _smart_string_alloc(smart_string *str, size_t len)
 		str->c = erealloc2(str->c, str->a + 1, str->len);
 	}
 }
+
+ZEND_API void ZEND_FASTCALL smart_str_append_escaped_truncated(smart_str *str, zend_string *value, size_t length)
+{
+	smart_str_append_escaped(str, ZSTR_VAL(value), MIN(length, ZSTR_LEN(value)));
+
+	if (ZSTR_LEN(value) > length) {
+		smart_str_appendl(str, "...", sizeof("...")-1);
+	}
+}
+
+ZEND_API void ZEND_FASTCALL smart_str_append_scalar(smart_str *dest, zval *value, size_t truncate) {
+	ZEND_ASSERT(Z_TYPE_P(value) <= IS_STRING);
+
+	if (Z_TYPE_P(value) == IS_UNDEF || Z_TYPE_P(value) == IS_NULL) {
+		smart_str_appendl(dest, "NULL", sizeof("NULL")-1);
+	} else if (Z_TYPE_P(value) == IS_TRUE || Z_TYPE_P(value) == IS_FALSE) {
+		smart_str_appends(dest, Z_TYPE_P(value) == IS_TRUE ? "true" : "false");
+	} else if (Z_TYPE_P(value) == IS_DOUBLE) {
+		smart_str_append_printf(dest, "%.*G", (int) EG(precision), Z_DVAL_P(value));
```

Comment:
We should include something like https://github.com/php/php-src/blob/7c9ac479eb6fef9d1330cb32e3b46404aa578acc/ext/standard/var.c#L519-L521 here to make sure it looks like a float.

Reply:
Also shouldn't this be `%.*H`?

Reply:
Looks like we forgot to switch some uses of `%.*G` over to `%.*H` in PHP 8.0.

Reply:
Fixed this one, I can search for others and follow up later on.

There's no max dbl length defined for /Zend, I'm not sure if we should move the php def to /Zend ?

I've used 1000 for now, which is fallback in php.

Reply:
These prints are checked, can't use H.